### PR TITLE
Fixes #16982 - CVE 2016-7078 - Scope properly when no taxonomies are set

### PR DESCRIPTION
**Pull request by dLobatog**
The default scope for hosts and other objects did not restrict
properly by taxonomies. An user without organizations or
locations, could do anything it's permissions allow to.
The list of hosts was unrestricted and showed hosts in
any location or organization.

This is fixed to work so that:

Users without taxonomies, when set to 'any context' cannot see
anything, only non-taxable objects.

Users with taxonomies, when set to 'any context' can see
everything within all of their taxonomies context (including
children taxonomies).

Admins set to 'any context' can see everything, regardless
of whether it has a taxonomy or not.

Users or admins set to some organization/location scope
can only see stuff within scope.

---

Problems with this change: a LOT of our tests depend on 'no taxonomies' behavior being broken as it used to be. There are still some tests that need changing. If you feel like the changes are not so many, feel free to edit this PR (it's set to allow edits from maintainers). I'm leaving for a week since now so I doubt I'll have time to look at this PR.

If you think the approach is wrong feel free to solve in any other way. I think the good thing about this is that it'll force us to be more explicit about taxonomies in tests.

@jlsherrill @ehelms @thomasmckay (can't mention @katello team in here :/) this is likely to affect Katello tests a lot as well.

### CODE REVIEW

**Comment by dLobatog** [comment]
Done with the Foreman part - I'll try to fix Katello tests as well, help appreciated though.

File: app/models/concerns/taxonomix.rb
Comment (tbrisker) [inline app/models/concerns/taxonomix.rb]: Any reason to use `where('1=1')` rather then `where(nil)`?

Comment (dLobatog) [reply app/models/concerns/taxonomix.rb]: None, just that I thought it'd be easier to read

**Comment by dLobatog** [comment]
You can find the Katello part of the story in https://github.com/Katello/katello/pull/6447 - just a few tests were relying on non-admin users able to see 'unscoped' objects.

File: app/models/concerns/taxonomix.rb
Comment (tbrisker) [inline app/models/concerns/taxonomix.rb]: Would it make sense to create the shortcut for admin in any context earlier? 
i.e. first line in the function, something like `return nil if loc.nil? && org.nil? && User.current.try(:admin?)` so it is clear `nil` is only returned in one specific case?  (which might also make sense to comment that return nil == everything)
Then we could also simplify the logic, just have an array `inner_ids` that we push into and return. Let me know if i'm not clear.

File: test/controllers/api/v2/config_reports_controller_test.rb
Comment (tbrisker) [inline test/controllers/api/v2/config_reports_controller_test.rb:51]: is this (and similar changes in other tests) related?

File: test/controllers/subnets_controller_test.rb
Comment (tbrisker) [inline test/controllers/subnets_controller_test.rb]: well, did you find?

**Review by tbrisker — CHANGES_REQUESTED** [review]
In general looks good, added a few inline comments. Also, how about adding some more tests on the changed taxonomix functions to make sure we don't break this again?

File: test/controllers/subnets_controller_test.rb
Comment (dLobatog) [reply test/controllers/subnets_controller_test.rb]: Ah forget about that, it was a reminder to myself

File: test/controllers/api/v2/config_reports_controller_test.rb
Comment (dLobatog) [reply test/controllers/api/v2/config_reports_controller_test.rb:51]: Yeah, without it `auth_smart_proxies` in the controller will find other proxies with the feature that have 'else.where' in the name, and those might not necessarily have the Puppet feature

**Comment by dLobatog** [comment]
Latest update only puts the 'return nil if admin' condition at the beginning of the `taxable_ids` method and changes `where 1=1` to `where nil`.

**Comment by tbrisker** [comment]
I created https://github.com/dLobatog/foreman/pull/29 to explain what I meant for simplifying `taxable_ids`. Also, I'm still missing some new tests to make sure this method doesn't break again.

**Comment by dLobatog** [comment]
@tbrisker Thanks for the PR, I've tested it and I don't think it's equivalent in functionality when you have both organizations and locations enabled: https://github.com/dLobatog/foreman/pull/29/files#r87873373

As for testing - as you can see a long list of tests will break if this breaks again, but I'll include a few more that specifically test returning nil when no orgs/locs are selected (and taxable_ids is called from admin) and for returning User.current along the list of users.

**Comment by dLobatog** [comment]
Updated with extra tests!

File: test/models/taxonomix_test.rb
Comment (tbrisker) [inline test/models/taxonomix_test.rb:320]: Shouldn't we also test that for a non-admin user empty taxonomy scope returns nothing?

Comment (dLobatog) [reply test/models/taxonomix_test.rb:320]: Will do in a moment

**Comment by dLobatog** [comment]
@tbrisker Updated - the test ('can work with array of taxonomies') covers the case for checking the visibility of stuff that's in the current organization but not in the current location. I wrote 'returns no IDs for non-admin user' to check that it doesn't return all objects for a regular user without taxonomies.

**Review by tbrisker — APPROVED** [review]

**Comment by tbrisker** [comment]
LGTM, but I would be happy if another maintainer who is familiar with taxonomix/permissions could take a quick look just to be sure. Also, do we need the katello changes to get merged before or after this?

**Comment by dLobatog** [comment]
Katello stuff should be merged after

**Comment by dLobatog** [comment]
Repushed because a test from e18d69e89b544288d90926011d7bb8b1d46a182b was relying on the CVE to work - it's now fixed on `test/controllers/puppetclasses_controller_test.rb`

File: app/models/concerns/taxonomix.rb
Comment (ares) [inline app/models/concerns/taxonomix.rb]: calling `uniq` might remove some dupes, oth SQL can deal with it too

**Comment by ares** [comment]
visually looks good, I'd like to give it a try tomorrow, this should be covered in released notes since it will have impact on users that only enabled orgs/locs but don't use them, regular users won't see anything after upgrade

**Comment by ares** [comment]
> Users without taxonomies, when set to 'any context' cannot see
anything (at all)

I don't think this is entirely correct. User in any context can see resources that are not assigned to any orgnaization/location or are not even taxable. E.g. such user can see all unassigned provisioning templates if he or she has permission. That I could understand. With this patch applied it seems that user can see even more, e.g. hostgroup that is assigned to org A (user has no org/loc assigned).

Also it seems weird that with this patch, user in any context doesn't see resources from a specific taxonomies which he's assigned to. I have host group in org A, user is in org A, C (there's also B). In any context he does not see the host group, after choosing org A, the user can see it.

**Comment by dLobatog** [comment]
> Also it seems weird that with this patch, user in any context doesn't see resources from a specific taxonomies which he's assigned to. I have host group in org A, user is in org A, C (there's also B). In any context he does not see the host group, after choosing org A, the user can see it.

@ares Have you tried that with the latest version of this PR? I tried the following to reproduce that:

* User "robotest" in Location 'Spain', Organization 'Lobacorp" , Role 'viewer' - can see all Hostgroups in those taxonomies when visiting /hostgroups ("Any context" selected) - ✔️ 
* Create a Hostgroup in location 'Czech Republic', organization 'E-Corp' - it can't be seen from user "robotest" - ✔️ 
* Add Location "Czech Republic" and Organization "E-Corp" to User "robotest" . Refresh the page ("Any context" selected) - hostgroups in both Czech/Spain Lobacorp/E-Corp are displayed ✔️ 

All of these sound reasonable to me, is anything wrong there?

About the former part "User without taxonomies in any context can see resources that are not assigned to any orgnaization/location or are not even taxable". I partly agree:

1. Things that are not taxable are definitely visible to users without taxonomies. I do agree with that.
2. Things that are taxable, but have *no taxonomies assigned*, should not be visible to any users other than Admin. I don't agree with what you said that "no taxonomies assigned" means "visible to users without taxonomies" AND "not visible to users with taxonomies". 

IIRC the convention is that when taxonomies are enabled, "no taxonomies assigned" means "admins only". I think there was an old CVE that had to do with this precisely, I'll look it up.

PS: I found yet another security issue when testing all of the above 😞 I'll ping you to check foreman-security in a few hours

**Comment by dLobatog** [comment]
Hmmm found a bug regarding users with Orgs but not Locs or viceversa. Feel free to not review until that's fixed.

**Comment by ares** [comment]
> @ares Have you tried that with the latest version of this PR?

yes

> Things that are taxable, but have no taxonomies assigned, should not be visible to any users other than Admin. I don't agree with what you said that "no taxonomies assigned" means "visible to users without taxonomies" AND "not visible to users with taxonomies".

I also think such users should not see such resources. The patch does not seem to change the fact that they do though. I think it was never meant to work like this and no one thought too much about non-admin users with no taxonomies. I wonder why someone would create a user without taxonomies, only to manage non-taxable resources? I think it would be better to forbid such users and require some orgs/locs if they are enabled.

**Comment by dLobatog** [comment]
https://github.com/theforeman/foreman/pull/4070 - completes this basically, although some tests need to be fixed from a recent update

**Comment by ares** [comment]
Sorry, it needs rebasing, also a lot of tests seem to fail with the latest version. I forgot to add results of our conversation here but I think we agreed also on adding a migration to create "Global Organization" for resources and users that are unassigned and adding a validation so that non-user can't be saved without any taxonomies. Or is that meant for a separate PR?

**Comment by dLobatog** [comment]
@ares Seems like a separate issue, so separate PR (It'd be easier to review, this one is huge for a 'minor' change) - I already submitted this fix we talked about yesterday separately https://github.com/theforeman/foreman/pull/4070

I'll fix the new broken tests which I think stem from this line I added yesterday to fix the bug you mentioned in your previous comment https://github.com/theforeman/foreman/pull/3961/files#diff-501156756cdcbc510254e30f9e2a29daR102

**Comment by ares** [comment]
Ok, I'm fine with separate PR but I'd like to get it in the same Foreman release and now when 1.14 is close I'm not sure whether it's good idea to merge it without? Even though I'd love to get this into 1.14.

**Comment by tbrisker** [comment]
@dLobatog - this needs rebasing
@ares  - is there any reason not to merge after rebase?

**Comment by ares** [comment]
Nope, the only concern I had was that both PRs should get in to the same release. The second PR is doing something else that we discussed here but that does not block this one...

**Comment by dLobatog** [comment]
https://github.com/theforeman/foreman_remote_execution/pull/217 fixes related tests in REX.

**Comment by dLobatog** [comment]
WoC until I fix the recent test breakages, this PR is quite tedious to maintain as new behavior/tests can easily break it.

**Comment by dLobatog** [comment]
WoC until I fix the recent test breakages, this PR is quite tedious to maintain as new behavior/tests can easily break it.

**Comment by dLobatog** [comment]
Seems like tests are 'randomly' failing - will check out why.

**Comment by dLobatog** [comment]
Hound went crazy asking for double quotes everywhere, I removed the comments and fixed that on another PR

File: app/models/concerns/taxonomix.rb
Comment (tbrisker) [inline app/models/concerns/taxonomix.rb]: This will fail if only one taxonomy type is enabled as it will try to do `[...] & nil`. 
Also, using inner_ids for both a variable and a method is confusing, how about just calling it `ids`?

How about going at this the other way - start with all the ids and remove those that don't match:
```ruby
ids = unscoped.pluck(:id)
ids &= inner_ids(loc, Location, inner_method) if SETTINGS[:locations_enabled]
ids &= inner_ids(org, Organization, inner_method) if SETTINGS[:organizations_enabled]
```

File: app/models/concerns/taxonomix.rb
Comment (tbrisker) [inline app/models/concerns/taxonomix.rb]: `any_context?(taxonomy)==true` here always, because if `taxonomy.present?` we already returned. Therefore, we could also short-circuit and return `[]` if no user found and get rid of this whole if block

**Comment by dLobatog** [comment]
Latest update just fixes a couple of tests in api/v2/hosts_controller and seeds

File: app/models/concerns/taxonomix.rb
Comment (tbrisker) [inline app/models/concerns/taxonomix.rb]: This should also include the child taxonomies, and could be improved:
```ruby
taxonomy_relation = taxonomy_class.to_s.underscore.pluralize
taxonomy_ids = User.current.taxonomy_and_child_ids(:"#{taxonomy_relation}")
return self.unscoped.joins(taxonomy_join_table).where('#{taxonomy_join_table}.taxonomy_id' => taxonomy_ids).pluck(:id)
```

**Review by tbrisker — CHANGES_REQUESTED** [review]
I think I found a couple of bugs and some improvement suggestions inline.

**Comment by dLobatog** [comment]
Seems like tests might randomly fail - http://ci.theforeman.org/job/test_develop_pr_core/11900/#showFailuresLink - will investigate.

Just saw those failures on https://github.com/theforeman/foreman_ansible/pull/74 without this PR being merged.. so I guess this happens on the current develop.

**Comment by dLobatog** [comment]
[test] ok to test

**Comment by dLobatog** [comment]
@tbrisker I implemented your suggestions here

**Comment by dLobatog** [comment]
@tbrisker Updated to solve the conflicts - hope no tests break again (only one broke in auth_source_ldap), this PR becomes just bigger the more tests rely on buggy behavior.

File: test/controllers/hosts_controller_test.rb
Comment (tbrisker) [inline test/controllers/hosts_controller_test.rb]: shouldn't this be `assert_equal`?

File: app/models/concerns/taxonomix.rb
Comment (tbrisker) [inline app/models/concerns/taxonomix.rb]: not necessary, `ids` will be an array for certain. (even if one of the previous lines makes it empty)

File: test/controllers/subnets_controller_test.rb
Comment (tbrisker) [reply test/controllers/subnets_controller_test.rb]: looks like this is still here

File: test/unit/tasks/seeds_test.rb
Comment (tbrisker) [inline test/unit/tasks/seeds_test.rb]: double unscoped

**Review by tbrisker — CHANGES_REQUESTED** [review]
Just a few minor comments inline. Should the katello PR be merged before or after this one?

**Comment by dLobatog** [comment]
@tbrisker I think this one should be merged before katello, otherwise it will not pass some of the tests there. I've updated this one with the changes you've commented on - thanks for reviewing!

**Comment by tbrisker** [comment]
@dLobatog I'm sorry, but looks like another test broke...

File: test/controllers/hosts_controller_test.rb
Comment (dLobatog) [inline test/controllers/hosts_controller_test.rb:459]: This is the test that broke in the last update, `.unscoped` was missing and also I compare now with `_id` instead of calling `.reload`.

**Comment by dLobatog** [comment]
@tbrisker Updated - left a comment inline about the test that broke

**Comment by tbrisker** [comment]
@dLobatog  - I'm going to merge this now so we have time to get any issues this may cause fixed in time for 1.15. Please make sure that the katello pr also gets merged asap.

File: db/seeds.d/07-provisioning_templates.rb
Comment (lzap) [inline db/seeds.d/07-provisioning_templates.rb:84]: Was this removed on purpose? This snippet is in use. @dLobatog

**Comment by lzap** [comment]
@dLobatog please ping Discovery maintainers as well with such a breaking changes :-) Allright let's get hands dirty.

**Pull request merged**
